aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -133,94 +139,89 @@ public void Configure(IApplicationBuilder app, IOptions<MyOptions> options)
+
+For more information, see <xref:fundamentals/middleware/write#per-request-middleware-dependencies>.
+
+Do ***not*** resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests. It's fine to:
+
+* Resolve a singleton service from a scoped or transient service.
+* Resolve a scoped service from another scoped or transient service.
+
+By default, in the development environment, resolving a service from another service with longer lifetime throws an exception. For more information, see [Scope validation](#sv).
 
 ### Singleton
 
-Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created the first time they're requested (or when `Startup.ConfigureServices` is run and an instance is specified with the service registration). Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide user code to manage the object's lifetime in the class.
+Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created either:
 
-In apps that process requests, singleton services are disposed when the <xref:Microsoft.Extensions.DependencyInjection.ServiceProvider> is disposed at app shutdown.
+* The first time they're requested.
+* When `Startup.ConfigureServices` is run and an instance is specified with the service registration.
+
+Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide code to dispose of the singleton. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.
+
+Singleton services must be thread safe and are often used in stateless services.
+
+In apps that process requests, singleton services are disposed when the <xref:Microsoft.Extensions.DependencyInjection.ServiceProvider> is disposed on application shutdown. Because memory is not released until the app is shut down, memory use with a singleton must be considered.
 
 > [!WARNING]
-> It's dangerous to resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests.
+> Do ***not*** resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests. It's fine to resolve a singleton service from a scoped or transient service.
 
 ## Service registration methods
 
 Service registration extension methods offer overloads that are useful in specific scenarios.
```

Comment:
Should it be mentioned here that IServiceCollection is a collection of ServiceDescriptor and an example of adding it like that explicitly (as below)? And state the extension methods makes it easier to add registrations.
`
var item = new ServiceDescriptor(typeof(IMyDependency), context = >new MyDependency(),   ServiceLifetime.Transient);

services.Add(item);
`

Reply:
@rahulpnath  great suggestion. Check the updated doc build.

Reply:
lgtm :)